README.md
## What is Google Maps Scraper what does it do?

This Google Maps scraper lets you get more and faster data from Google Places than the official [Google Places API](https://developers.google.com/places/web-service/search). Our unofficial Google Maps API enables you to extract all of the following data from Google Maps:

🔗   Title, subtitle, category, place ID, and URL

📍   Address, location, plus code and exact coordinates

☎️   Phone and website, if available

🏷   Menu and price, if available

🔒   Temporarily or permanently closed status

⌚️  Popular times - histogram & live occupancy

⭐️   Average rating (`totalScore`), review count, and review distribution

🍔   List of images (optional)

➕   List of detailed characteristics (`additionalInfo`, optional)

🧑‍🍳   Opening hours (optional)

🔍   People also search (optional)

The scraper also supports the scraping of all **detailed information about reviews**:

✅   Review text

✅   Published date

✅   Stars

✅   Review ID & URL

✅   Response from owner - text and published date

Personal data extraction about reviewers has to be **explicitly** enabled in input (see  [Personal data section](https://apify.com/drobnikj/crawler-google-places#personal-data)):

-   Reviewer name
-   Reviewer ID & URL
-   Reviewer number of reviews
-   Is Local Guide

Google Maps Scraper also provides **other handy features**:

🗺   Define search area - allows you to define the geographical area to scrape to a country, state, county, city, or postal code, thus speeding up the search (integration with Nomatim Maps API)

🔍   Automatic zooming - ensures maximum results

🌏   Language & translation settings

🗂   Reviews sorting

🛡   Proxy configuration

💻   Browser & scraping configuration


## How does  Google Maps Scraper work?

It works exactly as if you were searching through Google Maps and copying information from each page you find. It opens  [Google Maps website](https://www.google.com/maps/), goes to a specified location, then writes your search query into the search bar. Then it presses the *Next page* button until it reaches the final page or  `maxCrawledPlaces`. It enqueues all the places as separate pages and then copypastes all visible data into an organized document. To understand the process fully, just try it out in your browser - the scraper does exactly the same thing, only much faster.

## How much will scraping Google Maps cost?

Apify provides you with $5 free usage credits to use every month on the Apify Free plan, and you could get **up to 2,000 reviews from this Google Maps Scraper** for those credits. But the number of platform credits you use **depends on the complexity of your search**. 

If you need to get more data regularly, you should grab an Apify subscription. We recommend our $49/month Personal plan - you can get **up to 20,000 Google Maps results every month** with the free $49 in monthly usage credits from that plan! For more details about usage credits, see this video guide on ▷ [how to choose the right subscription plan](https://www.youtube.com/watch?v=s_89WpOsKRI).

For more details about Google Maps scraper usage, see the [Cost of usage tab](https://apify.com/drobnikj/crawler-google-places/cost-of-usage#features). 
<div align="center"><a href="https://console.apify.com/actors/nwua9Gu5YrADL7ZDj?asrc=blog" target="_blank">
<img width=20%" src="https://imgur.com/FS6hnof.png" /></a></div><br>   

## What are the advantages over the Google Maps API?

With the Google Maps API, you get $200 worth of credit usage every month free of charge. That means 28,500 maploads per month. However, the **Google Maps API caps your search results to 60**, regardless of the radius you specify. So, if you want to scrape data for bars in New York, for example, you'll get results for only 60 of the thousands of bars in the area. 

**Google Maps Scraper imposes no rate limits or quotas** and provides more cost-effective, comprehensive results, and also scrapes histograms for popular times, which aren't available in the official API.


## Is it legal to scrape Google Maps?

Web scraping is legal if you are extracting publicly available data which is most data on Google Maps. However, you should respect boundaries such as personal data and intellectual property regulations. You should only scrape personal data if you have a legitimate reason to do so, and you should also factor in Google's [Terms of Use](https://policies.google.com/terms?hl=en).


## How do I use Google Maps Scraper?

To understand how to configure and run the scraper, follow our step-by-step guide on  [how to scrape Google Maps](https://blog.apify.com/step-by-step-guide-to-scraping-google-maps/) or [watch a short video tutorial](https://www.youtube.com/watch?v=Wzfo3qSSbtU)  &#9655; on YouTube. 

[![Apify - G Maps](https://img.youtube.com/vi/J43AX9wu-NI/0.jpg)](https://www.youtube.com/watch?v=J43AX9wu-NI)

## What can I use the extracted data from Google Maps for?

You can use the extracted data to:

👉🏽 create a potential customer base and prospection files

👉🏽 find new clients

👉🏽 generate leads

👉🏽 search and analyze businesses similar to yours

👉🏽 monitor brand sentiment and service quality, and identify fake reviews

👉🏽 find where to buy products

👉🏽 analyze geo-spatial data for scientific or engineering work

👉🏽 develop a working market strategy

For more ideas on how to use the extracted data, check out our [industries pages](https://apify.com/industries)  for concrete ways web scraping results are already being used across many projects and businesses of all types and sizes - in [travel and logistics](https://apify.com/industries/travel-and-logistics), for instance.


## Want more options?

### Google Maps Reviews Scraper ⭐️

If you only want to scrape reviews, this targeted data scraper is a great option. [Google Maps Reviews Scraper](https://apify.com/zenisjan/google-maps-reviews-scraper) extracts all reviews for a single place on Google Maps. 

All you need to do is enter the URL of the location you want to scrape, and you'll get a dataset of all reviews, which you can download for business analysis and market research.

### Easy Google Maps Scraper 🌎

If you want an easier but more limited Google Maps scraping tool, this is a very handy scraper if you're only looking to extract, say, 40 results rather than 400. [Google Maps Reviews Scraper](https://apify.com/zenisjan/google-maps-reviews-scraper) is set up and configured to make your scraping tasks super quick and easy.

### Google Places API Radar Search📍

To use this tool, you need to acquire your own Google API key, as you would use the official Google API. So, why use [Google Places API Radar Search](https://apify.com/alexey/google-maps-radar-search) instead of the API?

Nearby search with the official API returns only 60 results. Google Places API Radar Search overrides this limit and gets all places in a specified location.

### Gas Station Scraper ⛽️

Use [Gas Station Scraper](https://apify.com/natasha.lekh/gas-stations-scraper) to find the lowest gas prices and timestamps of price updates from gas stations in your area.
       

## Input

### Input example
The Google Maps Scraper has the following input options:

![Apify  -  G Maps  Scraper  input](https://i.imgur.com/AxyIJG2.png)

Or here's its equivalent in JSON:
```json
{
  "maxCrawledPlaces": 10,
  "language": "en",
  "exportPlaceUrls": false,
  "includeHistogram": false,
  "includeOpeningHours": false,
  "includePeopleAlsoSearch": false,
  "additionalInfo": false,
  "maxImages": 10,
  "maxReviews": 0,
  "scrapeReviewerName": true,
  "scrapeReviewerId": true,
  "scrapeReviewerUrl": true,
  "scrapeReviewId": true,
  "scrapeReviewUrl": true,
  "scrapeResponseFromOwnerText": true,
  "proxyConfig": {
    "useApifyProxy": true
  },
  "maxCrawledPlacesPerSearch": 10,
  "searchStringsArray": [
    "pet shelter in Prague"
  ],
  "oneReviewPerRow": false,
  "reviewsSort": "newest",
  "reviewsTranslation": "originalAndTranslated",
  "allPlacesNoSearchAction": ""
}
```
<div align="center"><a href="https://console.apify.com/actors/nwua9Gu5YrADL7ZDj?asrc=blog" target="_blank">
<img width=20%" src="https://imgur.com/FS6hnof.png" /></a></div><br> 

## Output

The output from Google Maps Scraper is stored in a dataset. After the run is finished, you can download the dataset in various data formats (JSON, CSV, XML, RSS, HTML Table).

### Output example

![Apify  -  G Maps  Scraper  input](https://i.imgur.com/RaeEvdb.png)

Or here's its equivalent in JSON:
``` json 
[{
  "title": "Pet Heroes shop",
  "subTitle": "Pet Heroes - pomáháme zvířatům",
  "price": null,
  "menu": null,
  "categoryName": "Non-governmental organization",
  "address": "V Horkách 4, 140 00 Praha 4-Nusle, Czechia",
  "locatedIn": null,
  "neighborhood": "V Horkách 4",
  "street": "V Horkách 4",
  "city": "Prague 4-Nusle",
  "postalCode": "140 00",
  "state": null,
  "countryCode": "CZ",
  "plusCode": "3C6W+MR Prague 4, Czechia",
  "website": "http://www.petheroes.cz/",
  "phone": "+420 725 723 947",
  "temporarilyClosed": false,
  "location": {
    "lat": 50.0616625,
    "lng": 14.4470878
  },
  "permanentlyClosed": false,
  "totalScore": 4.8,
  "isAdvertisement": false,
  "rank": 10,
  "placeId": "ChIJJXTKRpmVC0cREdfm_mnO1e8",
  "categories": [
    "Non-governmental organization"
  ],
  "cid": "17281946099747575569",
  "url": "https://www.google.com/maps/place/Pet+Heroes+shop/@50.0616625,14.4470878,17z/data=!3m1!4b1!4m5!3m4!1s0x470b959946ca7425:0xefd5ce69fee6d711!8m2!3d50.0616625!4d14.4470878?hl=en",
  "searchPageUrl": "https://www.google.com/maps/search/pet+shelter+in+Prague/@37.6,-95.665,4z?hl=en",
  "searchString": "pet shelter in Prague",
  "scrapedAt": "2022-07-28T12:41:14.352Z",
  "reviewsCount": 17,
  "reviewsDistribution": {
    "oneStar": 0,
    "twoStar": 1,
    "threeStar": 0,
    "fourStar": 0,
    "fiveStar": 16
  },
  "imageUrls": [
    "https://lh5.googleusercontent.com/p/AF1QipO5KXrDiwPABpf329swB-ZqgNNbwSXt2EXdcb1C=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipMsy8BCwkMD8WJCNr4Qce98Tt_lLN_l1Tt1NhvI=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipPZbx-sQF7GlQ3iHvPZ7vPJmwSg3GmTjZwjcsAJ=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipO_S2d-hodZ7P1Ua_OUFNr7FOWIm02GeBAmnK8f=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipOte-mza93xiqM60Hbm9wQaExi7J6gyLOo53K2r=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipOUpt1d4TDfdXK5owjTANfYQfLG1dvAJRnwI8kg=w1920-h1080-k-no"
  ],
  "reviews": [],
  "orderBy": []
},
{
  "title": "Sdružení na ochranu zvířat v krajní nouzi",
  "subTitle": null,
  "price": null,
  "menu": null,
  "categoryName": "Pet adoption service",
  "address": "Na Pláni 2006, 150 00 Praha 5, Czechia",
  "locatedIn": null,
  "neighborhood": "Na Pláni 2006",
  "street": "Na Pláni 2006",
  "city": "Prague 5",
  "postalCode": "150 00",
  "state": null,
  "countryCode": "CZ",
  "plusCode": "397W+49 Prague 5, Czechia",
  "website": "https://www.kocici-utulek.cz/",
  "phone": "+420 603 225 948",
  "temporarilyClosed": false,
  "location": {
    "lat": 50.062872,
    "lng": 14.3958755
  },
  "permanentlyClosed": false,
  "totalScore": 4.3,
  "isAdvertisement": false,
  "rank": 9,
  "placeId": "ChIJuxlGAU6UC0cRI_jkTCUIboA",
  "categories": [
    "Pet adoption service"
  ],
  "cid": "9254343240589834275",
  "url": "https://www.google.com/maps/place/Sdru%C5%BEen%C3%AD+na+ochranu+zv%C3%AD%C5%99at+v+krajn%C3%AD+nouzi/@50.062872,14.3958755,17z/data=!3m1!4b1!4m5!3m4!1s0x470b944e014619bb:0x806e08254ce4f823!8m2!3d50.0628787!4d14.3958708?hl=en",
  "searchPageUrl": "https://www.google.com/maps/search/pet+shelter+in+Prague/@37.6,-95.665,4z?hl=en",
  "searchString": "pet shelter in Prague",
  "scrapedAt": "2022-07-28T12:42:10.969Z",
  "reviewsCount": 40,
  "reviewsDistribution": {
    "oneStar": 6,
    "twoStar": 0,
    "threeStar": 1,
    "fourStar": 2,
    "fiveStar": 31
  },
  "imageUrls": [
    "https://lh5.googleusercontent.com/p/AF1QipOeEWgXD8Jjmj3DpIa7U9VeJ3E83xaRpefxbYZh=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipN2XNEQQrjtpMIHLe0WlJHYWd4nhniifUiy9BYq=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipOShH8UZgA-gtJlc83n2uBLhgkd5HRacPIOx_V6=w1920-h1080-k-no",
    "https://lh5.googleusercontent.com/p/AF1QipPrcwuF0i7y32PUwX-ff-jdkRovD7XQ6fmBWHmr=w1920-h1080-k-no"
  ],
  "reviews": [],
  "orderBy": []
},
```

## Tips and tricks: define search area

### Country, state, county, city, and postal code vs. latitude and longitude

You can speed up the search and consume fewer platform credits by reducing the radius of the geographical area to scrape. You can use any combination of the geolocation parameters: `country`, `state`, `county`,  `city` & `postalCode`.

Keep in mind that the first five fields and the coordinate options are mutually exclusive.

### Automatic zooming


CHANGELOG.md
# 2022-09-23
*Fixes*
- BREAKING CHANGE: Removed `maxCrawledPlaces` from input completely (use `maxCrawledPlacesPerSearch` instead)
- Fixed `maxCrawledPlacesPerSearch` leading to scraper being hang out in some cases

# 2022-09-06
*Fixes*
- Fixes unstable image extraction

# 2022-09-05
*Fixes*
- Final round of optimizations and fixes of the search process. The scraper is now probably the fastest is has ever been finally reaching about 100 places per 1 compute unit even with using geolocation.

# 2022-09-02
*Fixes*
- Several optimizations to speed up the search page (scrolling & enqueueing places)
- Fixed extraction of images

# 2022-08-16
*Fixes*
- Improve extraction of additional infos for hotels.

# 2022-08-15
*Fixes*
- Fixed actor sometimes finishing prematurely when there were still requests in the queue (caused by the new background enqueueing system)

# 2022-08-05
*Fixes*
- Fixed reviews duplications that sometimes happened.
- Fixed extraction of the temporarilyClosed field.

# 2022-08-03
*Fixes*
- Fixed reviews extraction. After Google's change, the scraper was giving only up to 10 reviews. Now it works fully again. `newest` doesn't sort properly though yet.

# 2022-07-21
*Fixes*
- Finish fast when less than 120 places are found on a page. Previous implementation waited several seconds extra.

# 2022-07-20
*Fixes*
- Search pages now use scrolling instead of pagination. This makes the crawling a little slower and reduces the maximum number of places per page from 400 to 120. Use geolocation with zoom to work around this reduction. We might increase the default zoom by 1. in the near future.

# 2022-05-19
*Features*
- Added `gasPrices` to output. Available only for gas stations in US to the best of our knowledge.

# 2022-05-02
*Fixes*
- subTitle extraction works now

# 2022-04-04
*Fixes*
- Blocked responses on the search page now properly retry the request (no more unhandled promise rejection)
- Smoother search page pagination
- More informative logs
- Fixed consent approval if browser crashes

# 2022-03-16
*Fixes*
- `maxCrawledPlaces` + `exportPlaceUrls` was giving inconsistent number of results.

# 2022-03-14
*Features*
- Added `allPlacesNoSearch` to input. This option allows you to scrape all places shown on the map without the need for any search term.
- Added `reviewsStartDate` to input to extract only reviews newer than this date.
- Added `radiusKm` to the `Point` type in `customGeolocation`

# 2022-03-04
*Improvement*
- `additionalInfo` extraction is faster now.
- `additionalInfo` extraction for hotels and similar categories is more complete now:
Data which is not displayed on the Google page but present in the Google response is also extracted.

# 2022-03-03
- Lowering the default zoom values. The past setup made the scraping too slow and costly. The new defaults will speed up the scraping a lot while missing only a few places. You can still manually override the `zoom` parameter. New default values are:
`country` or `state` -> 12
`county` -> 14
`city` -> 15
`postalCode` -> 16
no geolocation -> 12

# 2022-02-28
*Fixes*
- `location` extraction works in (almost) all cases now (search URLs and URLs with place IDs will always work).

# 2022-02-21
*Features*
- Added `oneReviewPerRow` to input to enable expanding reviews one per output row

# 2022-02-17
*Fixes*
- `openingHours` extraction works in almost all cases now (search URLs and URLs with place IDs will always work).

# 2022-01-12
- Start URLs now correctly work from uploaded CSV files or Google Sheets. It uses to trim part of the URL.

# 2022-01-11
- Changed `polygon` input field to `customGeolocation`
- Added deeper section into Reamde on how you can provide your own exact coordinates

# 2022-01-11
*Breaking changes*
We decided it is time to change several default parameters to make the user experience smoother. These changes should not have a big effect on currect users.

- `city` and other geolocation parameters will have preference over `lat` & `long` if both are used (*in 99% cases users want to use the automatic location splitting to get the most results which doesn't work with direct `lat` & `long`*)
- `zoom` will no longer have a default value 12. Instead, it will change based on geolocation type like this:

`country` or `state` -> 12
`county` -> 14
`city` -> 17
`postalCode` -> 18
no geolocation -> 12

Users will still be able to specify the zoom and override this behavior.

See [Readme](https://apify.com/drobnikj/crawler-google-places#automatic-zooming) for more details

## 2021-12-14
*Breaking change*
- `reviewsSort` is now set to `newest` by default. This is because some places don't yield all reviews on other sortings (we are not sure if this is a bug or silent block on Google's side)

## 2021-11-15
*Fixes*
- `exportPlaceUrls` now properly dedupes the URLs
- added `categories` fields listing all categories the place is listed in

## 2021-11-11
*Fixes*
- Fixed `additionalInfo` for hotels
- Fixed `exportPlaceUrls` not checking for correct geolocation

## 2021-11-09
*Fixes*
- `website` field now displays the full URL. This fixes issue of blank `facebook.com` links.

## 2021-11-05
*Fixes*
- Fixed new layout of `additionalInfo`

## 2021-11-03
*Fixes*
- Improved reliability of scraping place detail, reviews and images (improving scrolling and back button interaction)

## 2021-10-13
*Features*
- Added `menu` to output
- Added `price` to output

## 2021-10-07
*Fixes*
- Fixed `popularTimesHistogram` which caused crash on some pages

## 2021-09-27
*Fixes*
- Fixed image extraction & make it optional (it should not crash the whole scrape)

## 2021-09-15
*Fixes*
- Fixed `temporarilyClosed` and `permanentlyClosed`
- Added a step for normalizing input Start URLs because those with wrong format don't contain JSON data

## 2021-09-14
*Fixes*
- Fixed popular times live and histogram

## 2021-09-10
https://github.com/drobnikj/crawler-google-places/pull/185
https://github.com/drobnikj/crawler-google-places/issues/181

*Fixes*

src/utils.js
const Apify = require('apify');
const Puppeteer = require('puppeteer');

const { DEFAULT_TIMEOUT, PLACE_TITLE_SEL, BACK_BUTTON_SEL } = require('./consts');

const { utils } = Apify;
const { log } = utils;

/**
 * Wait until google map loader disappear
 * @param {Puppeteer.Page} page
 * @return {Promise<void>}
 */
module.exports.waitForGoogleMapLoader = async (page) => {
    if (await page.$('#searchbox')) {
        // @ts-ignore
        await page.waitForFunction(() => !document.querySelector('#searchbox')
            .classList.contains('loading'), { timeout: DEFAULT_TIMEOUT });
    }
    // 2019-05-19: New progress bar
    await page.waitForFunction(() => !document.querySelector('.loading-pane-section-loading'), { timeout: DEFAULT_TIMEOUT });
};

/** @param {number} float */
module.exports.fixFloatNumber = (float) => Number(float.toFixed(7));

/**
 * @param {Puppeteer.Page} page
 */
 module.exports.getScreenshotPinsFromExternalActor = async (page) => {
    const base64Image = await page.screenshot({ encoding: 'base64' });
    const ocrActorRun = await Apify.call('alexey/google-maps-pins-map-ocr', { base64Image, mapURL: page.url() }, { memoryMbytes: 256 });
    if (ocrActorRun?.status !== 'SUCCEEDED') {
        log.error('getScreenshotPinsFromExternalActor', ocrActorRun);
        return [];
    }
    const externalDataset = await Apify.openDataset(ocrActorRun.defaultDatasetId, { forceCloud: true });
    const externalTAData = await externalDataset.getData({ clean: true });
    log.info(`[OCR]: Found ${externalTAData.items.length} pin(s) by run ${ocrActorRun?.id} for ${page.url()}`);
    // recalculate coordinates to pin center (current pin is 20x20px)
    const positionsFromActor = externalTAData.items.map((/** @type { any } */coords) => {
        return {
            x: coords?.x + 10,
            y: coords?.y + 10,
        }
    });
    return positionsFromActor;
}

/**
 * TODO: Add more cases
 * @param {Puppeteer.Page} page
 * @param {{ enqueued: number }} pageStats
 * @param {Array<{ x: number, y: number }>} ocrCoordinates,
 */
module.exports.moveMouseThroughPage = async (page, pageStats, ocrCoordinates) => {
    const plannedMoves = ocrCoordinates || [];
    // If we do not have coordinates from OCR actor then fill in viewport
    const viewport = page.viewport();
    if (!ocrCoordinates?.length && viewport) {
        const { width, height } = viewport;
        // If you move with less granularity, places are missed
        for (let y = 0; y < height; y += 10) {
            for (let x = 0; x < width; x += 10) {
                plannedMoves.push({ x, y });
            }
        }
    }
    log.info(`[SEARCH]: Starting moving mouse over the map to gather all places. Will do ${plannedMoves.length} mouse moves. This might take a few minutes: ${page.url()}`);
    let done = 0;
    for (const { x, y } of plannedMoves) {
        if (done !== 0 && done % 500 === 0) {
            log.info(`[SEARCH]: Mouse moves still in progress: ${done}/${plannedMoves.length}. Enqueued so far: ${pageStats.enqueued} --- ${page.url()}`);
        }
        await page.mouse.move(x, y, { steps: 5 });
        // mouse trick for processing OCR, otherwise places might be missed because mouse moved too fast
        if (ocrCoordinates?.length) {
            // wait for place detection
            await Apify.utils.sleep(1000);
            // move a bit and wait again
            await page.mouse.move(x + 4, y + 4, { steps: 5 });
            await Apify.utils.sleep(1000);
            // go back to top left corner to hide popup with preview
            // otherwise it might overlap with places nearby and prevent them from detection
            await page.mouse.move(0, 0, { steps: 5 });
            await Apify.utils.sleep(1000);
        }
        done++;
    }
}

/**
 * Method scrolls page to xpos, ypos.
 * @param {Puppeteer.Page} page
 * @param {string} selectorToScroll
 * @param {number} scrollToHeight
 */
module.exports.scrollTo = async (page, selectorToScroll, scrollToHeight) => {
    try {
        await page.waitForSelector(selectorToScroll);
    } catch (e) {
        log.warning(`Could not find selector ${selectorToScroll} to scroll to - ${page.url()}`);
    }
    await page.evaluate((selector, height) => {
        const scrollable = document.querySelector(selector);
        scrollable.scrollTop = height;
    }, selectorToScroll, scrollToHeight);
};

/**
 *
 * @param {string} sheetUrl
 * @returns {string | null} downloadUrl
 */
const convertGoogleSheetsUrlToCsvDownload = (sheetUrl) => {
    const CSV_DOWNLOAD_SUFFIX = 'gviz/tq?tqx=out:csv';
    // The lazy (+?) regex is important because we don't want to capture other slashes
    const baseUrlMatches = sheetUrl.match(/^.*docs.google.com\/spreadsheets\/d\/.+?\//g);

    if (!baseUrlMatches || baseUrlMatches.length === 0) {
        log.error(`Invalid start url provided (${sheetUrl}).
        Google spreadsheet url must contain: docs.google.com/spreadsheets/d/{SPREADSHEET_ID}/`);
        return null;
    }

    const baseUrl = baseUrlMatches[0];
    const downloadRequestUrl = `${baseUrl}${CSV_DOWNLOAD_SUFFIX}`;

    log.info(`Converting Google Sheets URL to a standardized format. If this doesn't work, please create an issue`);
    log.info(`${sheetUrl} => ${downloadRequestUrl}`);

    return downloadRequestUrl;
}

/**
 *
 * @param {string} downloadUrl
 * @returns {Promise<any[]>}
 */
const fetchRowsFromCsvFile = async (downloadUrl) => {
    const { body } = await utils.requestAsBrowser({ url: downloadUrl});
    const rows = body.replace(/[";]/g, '').split('\n');

    return Array.from(new Set(rows));
}

/**
 *
 * @param {string} fileUrl
 * @returns {boolean}
 */
 const isGoogleSpreadsheetFile = (fileUrl) => {
    const googleSpreadsheetMatches = fileUrl.match(/docs.google.com\/spreadsheets/g);
    return googleSpreadsheetMatches?.length === 1;
};

/**
 *
 * @param {string} fileUrl
 */
const parseStartUrlsFromFile = async (fileUrl) => {
    /** @type {string[]} */
    let startUrls = [];

    if (isGoogleSpreadsheetFile(fileUrl)) {
        const dowloadUrl = convertGoogleSheetsUrlToCsvDownload(fileUrl);
        if (dowloadUrl) {
            startUrls = await fetchRowsFromCsvFile(dowloadUrl)
        } else {
            log.warning(`WRONG INPUT: Google Sheets URL cannot be converted to CSV. `)
        }
    } else {
        // We assume it is some text file
        startUrls = await fetchRowsFromCsvFile(fileUrl);
    }

    const trimmedStartUrls = startUrls.map((url) => url.trim());
    return trimmedStartUrls.filter((url) => url.length);
};

/**
 *
 * @param {any[]} startUrls
 * @returns { Promise<{ url: string, uniqueKey: string }[]> }
 */
module.exports.parseRequestsFromStartUrls = async (startUrls) => {
    /**
     * @type { { url: string, uniqueKey: string }[] }
     */
    let updatedStartUrls = [];

    /**
     * `uniqueKey` is specified explicitly for each request object
     * as SDK otherwise wrongly normalizes it
     */

    for (const request of startUrls) {
        if (typeof request === 'string') {
            updatedStartUrls.push({
                url: request,
                uniqueKey: request,
            });
        } else {
            const { url, requestsFromUrl } = request;
            if (requestsFromUrl) {
                const parsedStartUrls = await parseStartUrlsFromFile(requestsFromUrl);
                const parsedRequests = parsedStartUrls.map((url) => ({
                    url,
                    uniqueKey: url
                }));
                updatedStartUrls = updatedStartUrls.concat(parsedRequests);
            } else {
                updatedStartUrls.push({
                    ...request,
                    uniqueKey: url,
                });
            }
        }
    }

    return updatedStartUrls;
};

/** @param {string} url */
module.exports.parseZoomFromUrl = (url) => {
    const zoomMatch = url.match(/@[0-9.-]+,[0-9.-]+,([0-9.]+)z/);
    return zoomMatch ? Number(zoomMatch[1]) : null;
};

/**
 * Waits until a predicate (funcion that returns bool) returns true
 *
 * ```
 * let eventFired = false;
 * await waiter(() => eventFired, { timeout: 120000, pollInterval: 1000 })
 * // Something happening elsewhere that will set eventFired to true
 * ```
 *
 * @param {function} predicate
 * @param {object} [options]
 * @param {number} [options.timeout]
 * @param {number} [options.pollInterval]
 * @param {string} [options.timeoutErrorMeesage]
 * @param {string} [options.successMessage]
 */
const waiter = async (predicate, options = {}) => {
    const { timeout = 120000, pollInterval = 1000, timeoutErrorMeesage, successMessage } = options;
    const start = Date.now();
    for (;;) {
        if (await predicate()) {
            if (successMessage) {
                log.info(successMessage);
            }
            return;
        }
        const waitingFor = Date.now() - start;
        if (waitingFor > timeout) {
            throw new Error(timeoutErrorMeesage || `Timeout reached when waiting for predicate for ${waitingFor} ms`);
        }
        await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }
};
module.exports.waiter = waiter;

/**
 * Navigates back to the details page
 * either by clicking back button or reloading the main page
 *
 * @param {Puppeteer.Page} page
 * @param {string} pageLabel label for the current page for error messages
 * @param {string} placeUrl URL for hard reload
 */
module.exports.navigateBack = async (page, pageLabel, placeUrl) => {
    const title = await page.$(PLACE_TITLE_SEL);
    if (title) {
        log.info('[PLACE]: We are still on the details page -> no back navigation needed');
        return;
    }
    try {
        const backButtonPresent = async () => {
            const backButton = await page.$(BACK_BUTTON_SEL);
            return backButton != null;
        }
        await waiter(backButtonPresent, {
            timeout: 2000,
            pollInterval: 500,
            timeoutErrorMeesage: `Waiting for backButton on ${pageLabel} page ran into a timeout after 2s on URL: ${placeUrl}`,
        });
        const navigationSucceeded = async () => {
            const backButton = await page.$(BACK_BUTTON_SEL);
            if (backButton) {
                await backButton.evaluate((backButtonNode) => {
                    if (backButtonNode instanceof HTMLElement) {
                        backButtonNode.click();
                    }
                });
            }
            const title = await page.$(PLACE_TITLE_SEL);
            if (title) {
                return true;
            }
        }
        await waiter(navigationSucceeded, {
            // timeout: 10000,
            // pollInterval: 500,
            timeoutErrorMeesage: `Waiting for back navigation on ${pageLabel} page ran into a timeout after 10s on URL: ${placeUrl}`,
        });
    } catch (/** @type {any}*/ e) {
        // As a last resort, we just reload the main page
        log.warning(`${e.message} - will hard reload the place page instead`);
        try {
            await page.goto(placeUrl);
            await page.waitForSelector(PLACE_TITLE_SEL);
            await Apify.utils.puppeteer.injectJQuery(page);
        } catch (e) {
            throw 'Reloading the page to navigate back failed, retrying whole request';
        }
    }
}

/**
 * @param {Puppeteer.Page} page
 * @param {string} url
 * @param {boolean} persistCookiesPerSession
 * @param {Apify.Session | undefined} session
 */
module.exports.waitAndHandleConsentScreen = async (page, url, persistCookiesPerSession, session) => {
    // TODO: Test if the new consent screen works well!

    const predicate = async (shouldClick = false) => {
        // handling consent page (usually shows up on startup), handles non .com domains
        const consentButton = await page.$('[action^="https://consent.google"] button');
        if (consentButton) {
            if (shouldClick) {
                await Promise.all([
                    page.waitForNavigation({ timeout: 60000 }),
                    consentButton.click()
                ]);
            }
            return true;
        }
        // handling consent frame in maps
        // (this only happens rarely, but still happens)
        for (const frame of page.mainFrame().childFrames()) {
            if (frame.url().match(/consent\.google\.[a-z.]+/)) {
                if (shouldClick) {
                    await frame.click('#introAgreeButton');
                }
                return true;
            }
        }
    };

    /**
     * Puts the CONSENT Cookie into the session
     */
    const updateCookies = async () => {
        if (session) {
            const cookies = await page.cookies(url);
            // Without changing the domain, apify won't find the cookie later.
            // Changing the domain can duplicate cookies in the saved session state, so only the necessary cookie is saved here.
            if (cookies) {
                let consentCookie = cookies.filter(cookie => cookie.name=="CONSENT")[0];
                // overwrite the pending cookie to make sure, we don't set the pending cookie when Apify is fixed
                session.setPuppeteerCookies([{... consentCookie}], "https://www.google.com/");
                if (consentCookie) {
                    consentCookie.domain = "www.google.com"
                }
                session.setPuppeteerCookies([consentCookie], "https://www.google.com/");
            }
        } else {
            log.warning("Session is undefined -> consent screen cookies not saved")
        }
    }

    await waiter(predicate, {
        timeout: 60000,
        pollInterval: 500,
        timeoutErrorMeesage: `Waiting for consent screen timeouted after 60000ms on URL: ${url}`,
        successMessage: `Approved consent screen on URL: ${url}`,
    });
    await predicate(true);
    if (persistCookiesPerSession) {
        await updateCookies();
    }
};

//

/**
 * Only certain formats of place URLs will give the JSON with data
 * Examples in /samples/URLS-PLACE.js
 * https://www.google.com/maps/place/All+Bar+One+Waterloo/@51.5027459,-0.1196255,17z/data=!4m5!3m4!1s0x487604b8703e7371:0x4fa0bac2e4eea2de!8m2!3d51.5027724!4d-0.11729
 * https://www.google.com/maps/place/All+Bar+One+Waterloo/@51.5031352,-0.1219012,17z/data=!3m1!5s0x487604c79a2ef535:0x6b1752373d1ab417!4m12!1m6!3m5!1s0x487604b900d26973:0x4291f3172409ea92!2slastminute.com+London+Eye!8m2!3d51.5032973!4d-0.1195537!3m4!1s0x487604b8703e7371:0x4fa0bac2e4eea2de!8m2!3d51.5027724!4d-0.11729
 * @param {string} url
 * @returns {string}
 */
module.exports.normalizePlaceUrl = (url) => {
    const match = url.match(/(.+\/data=)(.+)/);
    if (!match) {
        log.warning(`Cannot normalize place URL, didn't find data param --- ${url}`);
        return url;
    }
    const [, basePart, dataPart] = match;
    const hashPairs = dataPart.match(/\ds0x[0-9a-z]+:0x[0-9a-z]+/g);
    if (!hashPairs || hashPairs.length === 0) {
        log.warning(`Cannot normalize place URL, didn't find hash pairs --- ${url}`);
        return url;
    }
    const lastHashPair = hashPairs[hashPairs.length - 1];
    const lastHashPairSuffixRegex = new RegExp(`${lastHashPair}.+`);
    const lastHashPairSuffixMatch = dataPart.match(lastHashPairSuffixRegex);
    if (!lastHashPairSuffixMatch) {
        log.warning(`Cannot normalize place URL, cannot find the suffix after hash pair --- ${url}`);
        return url;
    }
    const firstDataPart = '!4m5!3m4!';
    const normalized = `${basePart}${firstDataPart}${lastHashPairSuffixMatch}`;
    log.debug(`Normalized Start URL: ${url} => ${normalized}`);
    return normalized;
}

/** @param {string} googleResponseString */
module.exports.unstringifyGoogleXrhResponse = (googleResponseString) => {
    return JSON.parse(googleResponseString.replace(')]}\'', ''));
};

